### PR TITLE
Improve the module doc parser - subcategories/empty defaults

### LINKED ISSUE #10706: Improve the module doc parser - subcategories/empty defaults
Propose the following improvements to the [docs API/tooling](https://github.com/PX4/Firmware/blob/v1.8.0/src/platforms/px4_module.h) for the [module documentation](https://dev.px4.io/en/middleware/modules_main.html)

1. Support for sub-categories. 
   - This is intended for categories like rangefinders - would be hard to find among all the other drivers.
   - This might be defined like:
      ```
      PRINT_MODULE_USAGE_NAME("sf1xx", "driver");  // module name, category
      PRINT_MODULE_USAGE_SUBCATEGORY("distance_sensor")   //subcategory
      ```
    - subcategories would be defined in own files named *category_subcategory.md*
    - subcategories would be auto-linked from near top of category page
    - subcategory page title would be something like 
       ```
       # subcategory (category)
       ```
    - If we like the idea of subcategories but not the work, alternative is to just use existing category mechanism for each subcategory. This works, but would not be possible to do automatic linking and have parent category in title.
1. Support for default to be declared "not applicable" and not rendered
   - Currently on many APIs you have to declare a default value and this always gets rendered.
      ```
      void PRINT_MODULE_USAGE_PARAM_INT(char option_char, int default_val, int min_val, int max_val,
		const char *description, bool is_optional);
      ```
      We should be able to specify either some sort of nothing pointer or -1 to mean "not applicable". 
   - If the 'not applicable" value is set for `default_val` then the text and value for default value should not be rendered in the markdown
   - Ideally this same mechanism should also be allowed for max, min and other possibly not applicable values (though these are not rendered)

### PR DESCRIPTION
**Pull request by bkueng**
Fixes #10706

@hamishwillee this adds support for empty defaults and subcategories (we currently have module docs only for one distance driver). The subcategories need to be added to the TOC of the dev-guide as well.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Needs a rebase and has build failures.

**Comment by priseborough** [comment]
This breaks the build on px4_fmu targets with my Ubuntu16.04LTS build machine withtoolcahin installed using latest scripts.

In file included from ../../src/platforms/px4_module.h:44:0,
                 from ../../src/platforms/common/module.cpp:39:
../../src/platforms/px4_atomic.h: In instantiation of 'class px4::atomic<bool>':
../../src/platforms/px4_module.h:421:42:   required from here
../../src/platforms/px4_atomic.h:72:2: error: non-constant condition for static assertion
  static_assert(__atomic_always_lock_free(sizeof(T), 0), "atomic is not lock-free for the given type T");
  ^
compilation terminated due to -Wfatal-errors.

**Comment by bkueng** [comment]
@priseborough do you mean https://github.com/PX4/Firmware/pull/11328? Which toolchain version do you use?

**Comment by dagar** [comment]
If you look under artifacts in the Jenkins build "continuous-integration/jenkins/branch" you'll see the generated metadata from the PR. http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/module_docs_updates/5/artifacts

**Comment by priseborough** [comment]
Apologies, I meant https://github.com/PX4/Firmware/pull/11328. Toolchain is as installed following Linux install instructions as of 1 Feb 2019.

**Comment by bkueng** [comment]
@priseborough what's the output of `arm-none-eabi-gcc --version`? CI is on 7.2.1.

@hamishwillee did you have time to look at this?

File: src/platforms/px4_module.h [hunk 1]
```diff
@@ -472,6 +472,12 @@ __EXPORT void PRINT_MODULE_DESCRIPTION(const char *description);
  */
 __EXPORT void PRINT_MODULE_USAGE_NAME(const char *executable_name, const char *category);
 
+/**
+ * @brief Specify a subcategory.
```
Comment (hamishwillee) [inline src/platforms/px4_module.h]: Perhaps mention that subcategories are optional?

> @brief Specify a subcategory (optional).

**Comment by hamishwillee** [comment]
By code inspection looks good. I'll try it out this afternoon.

**Comment by hamishwillee** [comment]
Testing shows this does as I would hope - thanks! I don't have ability to give approving review though, so you'll need that from someone else.

File: src/platforms/px4_module.h (hunk 1 above)
Comment (bkueng) [reply src/platforms/px4_module.h]: Applied

**Pull request merged**